File: netlify/functions/interest.py
import json

def handler(event, context):
    headers = {
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Headers": "Content-Type",
        "Access-Control-Allow-Methods": "GET, POST, OPTIONS",
        "Content-Type": "application/json"
    }

    if event['httpMethod'] == 'OPTIONS':
        return {
            'statusCode': 200,
            'headers': headers,
            'body': ''
        }

    try:
        body = json.loads(event['body'])
        principal = float(body.get('principal', 0))
        rate = float(body.get('rate', 0))
        time = float(body.get('time', 0))
        type = body.get('type', 'simple') # simple or compound
        frequency = int(body.get('frequency', 1)) # times per year for compound

        if principal <= 0 or time <= 0:
             return {
                'statusCode': 400,
                'headers': headers,
                'body': json.dumps({'error': 'Principal and time must be greater than 0'})
            }

        r = rate / 100

        if type == 'simple':
            interest = principal * r * time
            total = principal + interest
        else:
            # Compound Interest: A = P(1 + r/n)^(nt)
            total = principal * (1 + r/frequency) ** (frequency * time)
            interest = total - principal

        return {
            'statusCode': 200,
            'headers': headers,
            'body': json.dumps({
                'interest_earned': round(interest, 2),
                'total_amount': round(total, 2)
            })
        }

    except Exception as e:
        return {
            'statusCode': 500,
            'headers': headers,
            'body': json.dumps({'error': str(e)})
        }

File: netlify/functions/interest.js
exports.handler = async function (event, context) {
    // Set CORS headers
    const headers = {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Headers': 'Content-Type',
        'Access-Control-Allow-Methods': 'POST, OPTIONS',
        'Content-Type': 'application/json'
    };

    // Handle preflight request
    if (event.httpMethod === 'OPTIONS') {
        return {
            statusCode: 200,
            headers,
            body: ''
        };
    }

    if (event.httpMethod !== 'POST') {
        return {
            statusCode: 405,
            headers,
            body: JSON.stringify({ error: 'Method not allowed' })
        };
    }

    try {
        const data = JSON.parse(event.body);
        const { principal, rate, time, type, frequency } = data;

        const p = parseFloat(principal);
        const r = parseFloat(rate);
        const t = parseFloat(time);

        let interest, totalAmount;

        if (type === 'compound') {
            // Compound interest: A = P(1 + r/n)^(nt)
            const frequencyMap = {
                'annually': 1,
                'semi-annually': 2,
                'quarterly': 4,
                'monthly': 12,
                'daily': 365
            };
            const n = frequencyMap[frequency] || 1;
            totalAmount = p * Math.pow(1 + (r / 100) / n, n * t);
            interest = totalAmount - p;
        } else {
            // Simple interest: I = P * R * T / 100
            interest = (p * r * t) / 100;
            totalAmount = p + interest;
        }

        return {
            statusCode: 200,
            headers,
            body: JSON.stringify({
                principal: p,
                rate: r,
                time: t,
                interest_earned: parseFloat(interest.toFixed(2)),
                total_amount: parseFloat(totalAmount.toFixed(2))
            })
        };
    } catch (error) {
        return {
            statusCode: 400,
            headers,
            body: JSON.stringify({ error: 'Invalid request data', details: error.message })
        };
    }
};
